File: frontend/types/bigram.ts
export type BigramStatistics = {
  total_pairs: number;
  unique_pairs: number;
};

export type RecordPairRequest = {
  prev: string;
  curr: string;
};

export type RecordPairResponse = {
  success: boolean;
  total_pairs: number;
  unique_pairs: number;
};

export type RerankItem = {
  word: string;
  frequency: number;
  score: number;
};

export type RerankRequest = {
  prev: string;
  candidates: [string, number][];
};

export type RerankResponse = {
  prev: string;
  candidates: RerankItem[];
};

export type StatisticsResponse = BigramStatistics;

File: backend/app/api/schemas/bigram.py
from pydantic import BaseModel, Field

class RecordPairRequest(BaseModel):
    prev: str = Field(..., min_length=1)
    curr: str = Field(..., min_length=1)

class RecordPairResponse(BaseModel):
    success: bool
    total_pairs: int
    unique_pairs: int

class RerankRequest(BaseModel):
    prev: str = Field(..., min_length=1)
    candidates: list[tuple[str, int]] = Field(..., max_length=50)

class RerankItem(BaseModel):
    word: str
    frequency: int
    score: float

class RerankResponse(BaseModel):
    prev: str
    candidates: list[RerankItem]

class StatisticsResponse(BaseModel):
    total_pairs: int
    unique_pairs: int